### ISSUE TITLE
[compiler-v2] Upgrading code generated by v1 with code generated by v2 leads to CONSTRAINT_NOT_SATISIFIED 

### ISSUE THREAD

**Issue by wrwg** [aptos-labs/aptos-core#13234]
From @magnum6actual :

FYI - running into an issue with compiler v2.  I've got some simple code I'm using for the tutorial I'm writing.  On local, It runs tests, compiles and publishes in v1.  It runs tests & compiles in v2.  However, publish in v2 returns this error:

```
package size 9675 bytes
{
  "Error": "Simulation failed with status: Transaction Executed and Committed with Error CONSTRAINT_NOT_SATISFIED\nExecution failed with status: metadata and code bundle mismatch: Unknown key found: [99, 111, 109, 112, 105, 108, 97, 116, 105, 111, 110, 95, 109, 101, 116, 97, 100, 97, 116, 97]"
}
```

I did a force restart of my local testnet and it published fine with v2. So, perhaps upgrade related?

**Comment by magnum6actual** [comment]
To add more details, the version I tried to upgrade from was originally published with CLI v 3.1.  When I tried to upgrade with CLI 3.3 and compiler v2 - that's where the error came from.

So you should be able to replicate the error by publishing a module with CLI 3.1 and then try to upgrade with 3.3, compiler v2 (which is a likely workflow for many ecosystem devs when they make the switch).

**Comment by rahxephon89** [comment]
@wrwg, the issue will happen when an older version of CLI starts a local testnet, which does not recognize the compilation metadata and raises the error here: https://github.com/aptos-labs/aptos-core/blob/afcf9fd10000791c05d576d87542490eba9eb5dd/aptos-move/framework/src/module_metadata.rs#L252 
So users need to use CLI 3.2 or later to test the V2.

**Comment by wrwg** [comment]
This would mean we can't deploy any v2 code on current testnet/devnet before 1.12 has rolled out. The new metadata should be just ignored by older node code. Didn't we put it into its own key/value pair instead of the existing metadata?

**Comment by rahxephon89** [comment]
> This would mean we can't deploy any v2 code on current testnet/devnet before 1.12 has rolled out. The new metadata should be just ignored by older node code. Didn't we put it into its own key/value pair instead of the existing metadata?

Yeah, v2 code can be deployed to devnet but not testnet yet. We did not extend the CompiledModule with new key/value paire but instead used the metadata with its own key/value pair so the old CLI/binary cannot recognize it.

**Comment by rahxephon89** [comment]
Close because the fix was landed on main and 1.12
